scalding-core/src/main/scala/com/twitter/scalding/typed/cascading_backend/CascadingBackend.scala

```diff
@@ -182,9 +183,35 @@ object CascadingBackend {
   }
   private[this] val cache = new CompilerCache
 
-  private def compile[T](mode: Mode): FunctionK[TypedPipe, CascadingPipe] =
-    Memoize.functionK[TypedPipe, CascadingPipe](
-      new Memoize.RecursiveK[TypedPipe, CascadingPipe] {
+  private def compile(mode: Mode): FunctionK[TypedPipe, CascadingPipe] =
+    new FunctionK[TypedPipe, CascadingPipe] {
+      private val cache = HCache.empty[TypedPipe, CascadingPipe]
+
+      override def toFunction[U]: TypedPipe[U] => CascadingPipe[U] = cached
+
+      private def cached[U]: TypedPipe[U] => CascadingPipe[U] = {
+        // Don't cache `CrossPipe`, but cache `left` and `right` side of it
+        case cp@CrossPipe(left, right) =>
+          notCached(excludes = Set(left, right))(cp)
+        // Don't cache `Fork` and `WithDescriptionTypedPipe`
+        // since if we do cache them `CrossPipe` will end up being cached as well
+        case tp@Fork(_) =>
+          transform(tp, this)
+        case tp@WithDescriptionTypedPipe(_, _) =>
+          transform(tp, this)
+        // Cache all other typed pipes
+        case tp =>
+          cache.getOrElseUpdate(tp, transform(tp, this))
+      }
+
+      private def notCached(excludes: Set[TypedPipe[_]]): FunctionK[TypedPipe, CascadingPipe] =
+        new FunctionK[TypedPipe, CascadingPipe] {
+          override def toFunction[U]: TypedPipe[U] => CascadingPipe[U] = { tp =>
+            if (excludes.contains(tp)) cached(tp) else transform(tp, this)
+          }
+        }
+
+      private val transform: RecursiveK[TypedPipe, CascadingPipe] = new RecursiveK[TypedPipe, CascadingPipe] {
```

Comment:
I’m worried this is causing nothing below here to be cached. Note the use of rec here used to cache. Now it has been disabled for everything in here.

I think this is a pretty large difference from the current behavior and basically only the outer most pipe will be cache. I don’t think that’s what we want.

Reply:
I guess code isn't 100% clear, but no - everything apart from `Cross` implementation (cross itself, and not `left` and `right` arguments to it), `Fork` and `WithDescriptionTypedPipe` will be cached.

`transform` here is pure implementation of cascading backend tranformation without any caching logic applied. On top of that caching logic applied (in `cached`) which is:
- don't cache `Fork` and `WithDescriptionTypedPipe`
- don't cache `Cross`, but cache `left` and `right` branches of it
- cache everything else

Reply:
yeah, I think adding comments will help. There are two layers of caching here that is confusing.

1. the first cache is the explicit HCache
2. the second cache is the RecursiveK which caches each call.

So, I think you are right, the transform is caching, but I think nested Cross *will* be cached by transform.

so `a.cross(b).cross(c)` I think will hit the inner cross inside transform, which will be cached.

Do you think that's not right?

Reply:
I might know where your confusion comes from: before this code was using `Memoize` over `RecursiveK`. Now this `transform` function isn't caching at all.

In the case you wrote - first cross will be not cached, then it will use `notCached(Set(a.cross(b)), c))` which leads to `cached(a.cross(b))` which leads to *not caching inner cross as well. I guess I should rename `.cached` to `.withCachePolicy` or something like this.

Reply:
okay, yeah. I was getting confused with `new RecursiveK` thinking it was caching. It isn't. I see it is not controlling how the recursion is done.

Instead, you call it with `transform(tp, this)` which recurses using the entire outer strategy.

Yeah, I now think I understand and think it is correct, but I think it is a bit subtle. Can we add a comment that describes and summarizes the points of confusion and clears them up. We are going be confused later:

1. why are we doing this? We are avoiding caching certain node types because they make bad plans that are usually worse.
2. how are we achieving this? We are wiring up this kind of co-recursive check so that the recursion has a branch and is sure not to cache certain types of nodes...

something like this.

Reply:
Removed `RecursiveK` usage. I think it's more straightforward and simpler this way. Also added comment explaining motivation for this logic.